### ISSUE TITLE
[Pytorch upstream] perf regression with triton "div_rn"

### ISSUE THREAD

**Issue by weishi-deng** [intel/intel-xpu-backend-for-triton#5314]
### Describe the issue

We observe70% perf regression in this softmax kernel due to the change from / to div_rn. We expected the same performance since CUDA performance is not affected.
Last known good:

[softmax-old.py](https://github.com/user-attachments/files/22917657/softmax-old.py)

regression:

[softmax-new.py](https://github.com/user-attachments/files/22917658/softmax-new.py)

reproduce:
python softmax_xxx.py -k

### Environment details

triton: pytorch_triton_xpu-3.5.0+git1b0418a9
GPU: BMG

**Comment by jianyizh** [comment]
div_rn should have better accuracy, so #5306 should be fixed first. On cuda, we do not observe difference on performance, but we see performance drop on many models on xpu.

**Comment by jianyizh** [comment]
igc version 2.18.5-1188~25.04
libze-intel-gpu1                                      25.35.35096.9-1\~25.04\~ppa3                 amd64        Intel(R) Graphics Compute Runtime for oneAPI Level Zero.

**Comment by jianyizh** [comment]
@vlad-penkin please help fix this issue first

**Comment by etaf** [comment]
@jianyizh @weishi-deng  I've filed a PR https://github.com/pytorch/pytorch/pull/165406 to fallback div_rn to "/" until #5306 and #5314 fixed.

**Comment by jianyizh** [comment]
@vlad-penkin We see similar issue for tl.sqrt_rn, so please check all related functions

**Comment by jianyizh** [comment]
reproducer for sqrt issue on PVC, you can replace libdevice.sqrt by tl.sqrt_rn. 
2.700ms    2.910GB    1077.72GB/s 
use tl.sqrt_rn will get:
1040.241ms    2.910GB    2.80GB/s

```python
import triton
import triton.language as tl

from torch._inductor.runtime import triton_helpers, triton_heuristics
from torch._inductor.runtime.triton_helpers import libdevice, math as tl_math
from torch._inductor.runtime.hints import AutotuneHint, ReductionHint, TileHint, DeviceProperties
triton_helpers.set_driver_to_gpu()

from torch._dynamo.testing import rand_strided
from torch._C import _xpu_getCurrentRawStream as get_raw_stream
import torch

@triton_heuristics.pointwise(
    size_hints={'x': 1073741824}, 
    filename=__file__,
    triton_meta={'signature': {'in_out_ptr0': '*fp16', 'in_ptr0': '*fp16', 'in_ptr1': '*fp16', 'in_ptr2': '*fp16', 'in_ptr3': '*fp16', 'xnumel': 'i32', 'XBLOCK': 'constexpr'}, 'device': DeviceProperties(type='xpu', index=0, multi_processor_count=64, cc={'architecture': 13136561920, 'device_id': 3029, 'driver_version': '1.6.33578+38', 'gpu_eu_count': 512, 'gpu_subslice_count': 64, 'has_atomic64': True, 'has_bfloat16_conversions': True, 'has_fp16': True, 'has_fp64': True, 'has_subgroup_2d_block_io': True, 'has_subgroup_matrix_multiply_accumulate': True, 'has_subgroup_matrix_multiply_accumulate_tensor_float32': False, 'max_compute_units': 512, 'max_num_sub_groups': 64, 'max_work_group_size': 1024, 'name': 'Intel(R) Data Center GPU Max 1550', 'platform_name': 'Intel(R) oneAPI Unified Runtime over Level-Zero', 'sub_group_sizes': [16, 32], 'total_memory': 68702699520, 'type': 'gpu', 'vendor': 'Intel(R) Corporation', 'version': '12.60.7'}, major=None, regs_per_multiprocessor=None, max_threads_per_multi_processor=None, warp_size=32), 'constants': {}, 'configs': [{(0,): [['tt.divisibility', 16]], (1,): [['tt.divisibility', 16]], (2,): [['tt.divisibility', 16]], (3,): [['tt.divisibility', 16]], (4,): [['tt.divisibility', 16]], (5,): [['tt.divisibility', 16]]}]},
    inductor_meta={'grid_type': 'Grid1D', 'autotune_hints': set(), 'kernel_name': 'triton_poi_fused__native_batch_norm_legit_no_training_relu_0', 'mutated_arg_names': ['in_out_ptr0'], 'optimize_mem': True, 'no_x_dim': False, 'num_load': 5, 'num_reduction': 0, 'backend_hash': '6848BC735BDC4BE5423D07D0CC808B4B31157262F796D51578CDC4523AB137C3', 'are_deterministic_algorithms_enabled': False, 'assert_indirect_indexing': True, 'autotune_local_cache': True, 'autotune_pointwise': True, 'autotune_remote_cache': None, 'force_disable_caches': True, 'dynamic_scale_rblock': True, 'max_autotune': False, 'max_autotune_pointwise': False, 'min_split_scan_rblock': 256, 'spill_threshold': 16, 'store_cubin': False, 'tiling_scores': {'x': 4364894464}, 'kernel_num_gb': 2.909929728, 'kernel_flop': 0},
    min_elem_per_thread=0
)
@triton.jit
def triton_poi_fused__native_batch_norm_legit_no_training_relu_0(in_out_ptr0, in_ptr0, in_ptr1, in_ptr2, in_ptr3, xnumel, XBLOCK : tl.constexpr):
    xnumel = 727482368
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = tl.full([XBLOCK], True, tl.int1)
    x2 = xindex
    x0 = (xindex % 32)
    tmp0 = tl.load(in_out_ptr0 + (x2), None).to(tl.float32)
    tmp2 = tl.load(in_ptr0 + (x0), None, eviction_policy='evict_last').to(tl.float32)
    tmp5 = tl.load(in_ptr1 + (x0), None, eviction_policy='evict_last').to(tl.float32)
    tmp15 = tl.load(in_ptr2 + (x0), None, eviction_policy='evict_last').to(tl.float32)
    tmp18 = tl.load(in_ptr3 + (x0), None, eviction_policy='evict_last').to(tl.float32)
    tmp1 = tmp0.to(tl.float32)
    tmp3 = tmp2.to(tl.float32)
    tmp4 = tmp1 - tmp3
    tmp6 = tmp5.to(tl.float32)
    tmp7 = 0.001
    tmp8 = tmp6 + tmp7
    tmp9 = libdevice.sqrt(tmp8)
    # tmp9 = tl.sqrt_rn(tmp8)
    tmp10 = tl.full([1], 1, tl.int32)
    tmp11 = (tmp10 / tmp9)
    tmp12 = 1.0
    tmp13 = tmp11 * tmp12
    tmp14 = tmp4 * tmp13
    tmp16 = tmp15.to(tl.float32)
    tmp17 = tmp14 * tmp16
    tmp19 = tmp18.to(tl.float32)
    tmp20 = tmp17 + tmp19
    tmp21 = tmp20.to(tl.float32)
    tmp22 = tl.full([1], 0, tl.int32)
    tmp23 = triton_helpers.maximum(tmp22, tmp21)
    tl.store(in_out_ptr0 + (x2), tmp23, None)


def get_args():
    arg_0 = rand_strided((1024, 32, 149, 149), (710432, 1, 4768, 32), device='xpu:0', dtype=torch.float16)
    arg_1 = rand_strided((32,), (1,), device='xpu:0', dtype=torch.float16)
    arg_2 = rand_strided((32,), (1,), device='xpu:0', dtype=torch.float16) +1.
    arg_3 = rand_strided((32,), (1,), device='xpu:0', dtype=torch.float16)
    arg_4 = rand_strided((32,), (1,), device='xpu:0', dtype=torch.float16)
    return arg_0, arg_1, arg_2, arg_3, arg_4, 727482368,


def call(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        stream0 = get_raw_stream(0)
        triton_poi_fused__native_batch_norm_legit_no_training_relu_0.run(*args, stream=stream0)


def benchmark_all_configs(args):
    with torch.xpu._DeviceGuard(0):
        torch.xpu.set_device(0)
        return triton_poi_fused__native_batch_norm_legit_no_training_relu_0.benchmark_all_configs(*args)


if __name__ == '__main__':
    from torch._inductor.runtime.benchmarking import benchmarker

    args = get_args()
    ms = benchmarker.benchmark_gpu(lambda: call(args), rep=40)
    num_gb = 2.909929728
    gb_per_s = num_gb / (ms / 1e3)
    print(f"{ms:.3f}ms    {num_gb:.3f}GB    {gb_per_s:.2f}GB/s")
```

**Comment by wdziurdz** [comment]
In the first (old) version, we use a standard division operation x / y, which compiles to %result = llvm.fdiv %x_27, %y_37 : f32
 
In the new softmax version, we use triton.language.div_rn(x, y), which compiles to %result = llvm.call spir_func @__imf_fdiv_rn(%x_27, %y_37) : (f32, f32) -> f32
 
It seems that the only difference between these two IRs is the external call to the SPIR-V function __imf_fdiv_rn instead of using llvm.fdiv.

I analyzed the code where PreciseDivFOp is replaced (since triton.language.div_rn is first converted to tt.precise_divf, which then maps to __imf_fdiv_rn) to __imf_fdiv_rn, and checked how this is handled by AMD and NVIDIA:

NVIDIA replaces it with an external call __nv_fdiv_rn and AMD replaces it with llvm::FDivOp (which is llvm.fdiv) which probably bug.

Additionally, I simplified the softmax kernel into small test kernels that only perform the division operation, and I still observe the same performance degradation. Scripts to reproduce and ir's in attachment. Only one difference it is call __imf_fdiv_rn.  
[div_old.py](https://github.com/user-attachments/files/23093597/div_old.py)
[div_new.py](https://github.com/user-attachments/files/23093598/div_new.py)
ir's:
[new_div.txt](https://github.com/user-attachments/files/23093625/new_div.txt)
[old_div.txt](https://github.com/user-attachments/files/23093624/old_div.txt)

Also, I profiled these small test kernels using the PyTorch profiler and also noticed that __imf_fdiv_rn is slower than llvm.fdiv. Screenshots are below:
<img width="1709" height="754" alt="Image" src="https://github.com/user-attachments/assets/1559fafb-6552-45ab-a2b4-3e8f3244c859" />
<img width="1598" height="744" alt="Image" src="https://github.com/user-attachments/assets/4285cbbd-5c36-48a1-9917-811ed4f88a89" />

To summarize, if this performance drop is not acceptable, I think someone from the IGC team should look into it, since the drop is isolated to the __imf_fdiv_rn call.

**Comment by HBN-MichalSzy** [comment]
@etiotto , @whitneywhtsang , @jianyizh please take a look at Witold's status and advise on next steps - can you take it up with IGC team?

@vlad-penkin fyi.

**Comment by whitneywhtsang** [comment]
`__imf_fdiv_rn` is more precise than `llvm.fdiv` (https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py#L1149), so it is not surprising that it takes longer. Please open an issue for DPC++ Runtime Library team to determine if the performance of `__imf_*_rn` builtins can be improved.

**Comment by etiotto** [comment]
Summary: This is not a bug in Triton nor it is a regression because the Triton kernel generated now contains  "triton.language.div_rn" rather than the standard "/" it contained before.  Triton correctly lowers "triton.language.div_rn"  to __imf_fdiv_rn

The strange thing is that "__imf_fdiv_rn" (Fast Division with Round to nearest) should not be slower that the standard division operation.  This should be investigated by the math library and IGC teams. @vlad-penkin can you open a Jira pls.

**Comment by whitneywhtsang** [comment]
Talked to Andrey Stepin, the reason why `__imf_fdiv_rn` is slow, is because it needs to emulate `__nv_fdiv_rn` to get the same results for all cases. We need to find a way to lower to precise div without using `__imf_fdiv_rn`.

**Comment by whitneywhtsang** [comment]
Looked into lowering to:
```
  %13 = fpext float %12 to double
  %14 = fpext float %10 to double
  %15 = tail call double @llvm.experimental.constrained.fdiv.f64(double %14, double %13, metadata !"round.tonearest", metadata !"fpexcept.strict") #4
  %16 = fptrunc double %15 to float

attributes #4 = { strictfp }
```
LLVM opt preserve it, but IGC opt makes it not precise. IGC only supports `FPRoundingMode` decoration for conversion instructions, but not for arithmetic. Tried to add FreezeOp and WaveBarrier, but both cannot stop the IGC optimization.

**Comment by whitneywhtsang** [comment]
With https://github.com/intel/intel-xpu-backend-for-triton/pull/5415, on PVC softmax-new.py, `9.576ms    1.895GB    197.93GB/s` -> `2.964ms    1.895GB    639.44GB/s`.

**Comment by etiotto** [comment]
Closing given that Triton has implemented a workaround now.
